enhancements/installer/aws-custom-edge-machineset-local-zones.md

```diff
@@ -0,0 +1,677 @@
+    spec:
+      metadata:
+        labels:
+          zone_type: local-zone
+          zone_group: us-east-1-nyc-1
+          node-role.kubernetes.io/edge: ""
+      taints:
+        - key: node-role.kubernetes.io/edge
+          effect: NoSchedule
+      providerSpec:
+        value:
+          ami:
+            id: ${AMI_ID}
+          apiVersion: amachine.openshift.io/v1beta1
+          blockDevices:
+          - ebs:
+              volumeSize: 120
+              volumeType: gp2
+          credentialsSecret:
+            name: aws-cloud-credentials
+          deviceIndex: 0
+          iamInstanceProfile:
+            id: ${CLUSTER_ID}-worker-profile
+          instanceType: c5d.2xlarge
+          kind: AWSMachineProviderConfig
+          placement:
+            availabilityZone: us-east-1-nyc-1a
+            region: us-east-1
+          securityGroups:
+          - filters:
+            - name: tag:Name
+              values:
+              - ${CLUSTER_ID}-worker-sg
+          subnet:
+          - filters:
+            - name: tag:Name
+              values:
+              - ${CLUSTER_ID}-public-us-east-1-nyc-1a  (1)
+          publicIp: true
+          tags:
+          - name: kubernetes.io/cluster/${CLUSTER_ID}
+            value: owned
+          userDataSecret:
+            name: worker-user-data
+```
+
+- 1) The `tag:Name` for the public subnet on the Local Zone location.
+
+The cloud compute resources will be created by installer:
+
+- 3 control plane nodes on the private subnets in the default Availability Zones (parent region)
+- 3 compute worker nodes on the private subnets in the default Availability Zones (parent region)
+- 1 compute edge node on the public subnet in Local Zone
+
+### Workflow Description
+
+TBD
+
+#### Variation [optional]
+
+<!-- If the cluster creator uses a standing desk, in step 1 above they can
+stand instead of sitting down.  
+
+See
+https://github.com/openshift/enhancements/blob/master/enhancements/workload-partitioning/management-workload-partitioning.md#high-level-end-to-end-workflow
+and https://github.com/openshift/enhancements/blob/master/enhancements/agent-installer/automated-workflow-for-agent-based-installer.md for more detailed examples. -->
+
+TBD
+
+### API Extensions
+
+Not applied. No new APIs will be added.
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Public subnet and outbound traffic
+
+It is a best practice to place the worker nodes in private subnets, although the
+AWS provides limited services in the Local Zones requiring a more complex setup to do so.
+
+Those are some options to create subnets on Local Zones:
+
+1) Create a subnet and associate it to a route table with the internet gateway as default.
+2) Create a private subnet, associating it to a route table from the zone in the
+  parent region. The cost will be twice as the traffic should cross the zone to
+  access the internet through a Nat Gateway, which also has fees to be applied.
```

Comment:
Have we measured the cost of adding an edge worker? Kubelet to Kube API traffic is pretty intense so I wonder what that's going to cost per node?

Reply:
The internal (within the VPC: kubelet<->NLBs) data transfer costs will be the same of the other (normal workers) nodes[1][2]. The compute resources in Local Zones is more expensive, for example `m5.xlarge` hosted in the parent zones of `us-west-2` will cost $0.192/hr, and in the LAX (Los Angeles) Local Zone is $0.23/hr.

The scenario described in this specific line is when following the architecture of using private subnets in Local Zone re-using Nat Gateway from the parent region, as NAT GW is not supported on the Local Zones locations.

In this case, the following fees will be applied when the node needs access to the internet:
- Fee for data transferred "in" to and "out" from NAT GW located in another AZ ($0.01/GB for each direction)
- NAT GW fee per GB processed $0.045

There is another cost of $0.045/hr for NAT GW allocation. But this is not exclusive to a node in LZ, as we can assume that the Nat GW is already created to be used in a valid AZ.

When the node is in the public subnet (described in this EP), the fees applied will be the same to access the internet.

[1] [EC2 Data transfer fees](https://aws.amazon.com/ec2/pricing/on-demand/)
[2] [EC2 Local Zones Pricing](https://aws.amazon.com/about-aws/global-infrastructure/localzones/pricing/):
> Data Transfer charges in AWS Local Zones is the same as in the Availability Zones today. Learn more about data transfer prices “in” and “out” of Amazon EC2 here.

Reply:
> So is this to say a local zone instance doesn't actually cost significantly more (aside from the instance type charges) than a regular az instance?

Yes, will not cost significantly more. The EC2 costs are fixed by location. In the case mentioned above, it was 17% more expensive.

> The data transfer costs aren't actually different?

No. The reference[2] from my last comment can confirm it. It was copied from the AWS Documentation page.

Reply:
Cool, this is useful context, can we make sure this is noted somewhere

Reply:
I created a [new section "Infrastructure costs"](https://github.com/openshift/enhancements/pull/1232/commits/fdbdba4c38f3b17c5159b4f5206a6a3671bc74ce#diff-f2590494c9138551ab5b3d5632d11170fe8742a128256f9e1d8ebc63baa19f41R620-R645) explaining what we discussed here. Thanks for the feedback.